// eslint-plugin-jsdoc/dist/rules/matchDescription.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _iterateJsdoc = _interopRequireDefault(require("../iterateJsdoc"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const tagsWithDescriptions = ['param', 'arg', 'argument', 'returns', 'return'];

var _default = (0, _iterateJsdoc.default)(({
  jsdoc,
  report,
  context,
  utils
}) => {
  const options = context.options[0] || {};

  const validateDescription = (description, tag) => {
    const regex = new RegExp((tag && typeof options.tags[tag] === 'string' ? options.tags[tag] : options.matchDescription // If supporting Node >= 10, we could loosen to this for the
    //   initial letter: \\p{Upper}
    ) || '^[A-Z`\\d_](?:[\\s\\S]*[.?!`])?$', 'u');

    if (!regex.test(description)) {
      report('JSDoc description does not satisfy the regex pattern.');
      return true;
    }

    return false;
  };

  if (jsdoc.description && validateDescription(jsdoc.description)) {
    return;
  }

  if (!options.tags || !Object.keys(options.tags).length) {
    return;
  }

  const tags = utils.filterTags(({
    tag
  }) => {
    return tagsWithDescriptions.includes(tag) && {}.hasOwnProperty.call(options.tags, tag) && options.tags[tag];
  });
  tags.some(tag => {
    const description = _lodash.default.trimStart(tag.description, '- ');

    return validateDescription(description, tag.tag);
  });
}, {
  meta: {
    schema: [{
      additionalProperties: false,
      properties: {
        matchDescription: {
          format: 'regex',
          type: 'string'
        },
        tags: {
          patternProperties: {
            '.*': {
              oneOf: [{
                format: 'regex',
                type: 'string'
              }, {
                enum: [true],
                type: 'boolean'
              }]
            }
          },
          type: 'object'
        }
      },
      type: 'object'
    }],
    type: 'suggestion'
  }
});

exports.default = _default;
module.exports = exports.default;
